File: services/api/src/routes/cultural.py
# ruff: noqa: B008
"""Cultural Notes API routes for the French Learning Platform.

Endpoints:
- GET  /notes              -- List cultural notes by CEFR level and category
- GET  /notes/{id}         -- Get full cultural note detail
- POST /notes/{id}/vocabulary/{vid}/add -- Add vocabulary to SRS from note
- POST /generate           -- Trigger async cultural content generation
"""

from __future__ import annotations

import logging
from datetime import UTC, datetime, timedelta
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from pydantic import BaseModel, Field
from services.api.src.middleware.auth import UserInfo, get_current_user
from services.shared.models.vocabulary import CEFRLevel
from services.shared.srs.fsrs import FSRSScheduler

logger = logging.getLogger(__name__)

router = APIRouter()

_fsrs = FSRSScheduler()

# ---------------------------------------------------------------------------
# Request / response schemas
# ---------------------------------------------------------------------------

VALID_CATEGORIES = frozenset(
    {"history", "neighborhoods", "etiquette", "cuisine", "daily_life"}
)


class CulturalNotePreview(BaseModel):
    """Preview of a cultural note for list views."""

    id: UUID
    cefr_level: str
    title_es: str
    title_fr: str
    category: str
    preview_es: str
    vocabulary_count: int
    reviewed: bool


class CulturalNoteListResponse(BaseModel):
    """Paginated list of cultural note previews."""

    notes: list[CulturalNotePreview]
    total: int


class VocabularyRef(BaseModel):
    """Vocabulary reference inside a cultural note."""

    id: UUID
    french_text: str
    spanish_translation: str
    in_user_review_queue: bool


class CulturalNoteDetail(BaseModel):
    """Full cultural note content."""

    id: UUID
    cefr_level: str
    title_es: str
    title_fr: str
    content_fr: str
    content_es: str
    vocabulary: list[VocabularyRef]
    category: str
    is_generated: bool
    reviewed: bool
    created_at: str


class AddVocabularyResponse(BaseModel):
    """Response after adding vocabulary to SRS."""

    vocabulary_item_id: UUID
    added_to_review: bool
    first_review_date: str


class GenerateRequest(BaseModel):
    """Request to generate a new cultural note."""

    cefr_level: CEFRLevel
    category: str = Field(
        description="Cultural category: history, neighborhoods, etiquette, cuisine, daily_life"
    )
    topic_hint: str | None = Field(
        default=None, description="Optional topic hint for generation"
    )
    align_with_vocabulary: list[UUID] = Field(
        default_factory=list,
        description="Vocabulary item IDs to align the article with",
    )


class GenerateResponse(BaseModel):
    """Response after triggering content generation."""

    generation_id: str
    status: str
    ai_platform: str
    estimated_completion_seconds: int


# ---------------------------------------------------------------------------
# Helper: get Supabase client from request
# ---------------------------------------------------------------------------


def _get_supabase(request: Request) -> Any:
    """Extract the Supabase client from app state."""
    supabase = getattr(request.app.state, "supabase", None)
    if supabase is None:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Database client not available.",
        )
    return supabase


# ---------------------------------------------------------------------------
# GET /notes -- List cultural notes
# ---------------------------------------------------------------------------


@router.get(
    "/notes",
    response_model=dict[str, CulturalNoteListResponse],
)
async def list_cultural_notes(
    request: Request,
    cefr_level: CEFRLevel = Query(
        ..., description="CEFR level filter"
    ),
    category: str | None = Query(
        default=None,
        description="Category filter: history, neighborhoods, etiquette, cuisine, daily_life",
    ),
    limit: int = Query(
        default=10, ge=1, le=50, description="Page size"
    ),
    offset: int = Query(
        default=0, ge=0, description="Pagination offset"
    ),
    user: UserInfo = Depends(get_current_user),
) -> dict[str, Any]:
    """List cultural notes filtered by CEFR level and optional category."""
    supabase = _get_supabase(request)

    if category and category not in VALID_CATEGORIES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                f"Invalid category '{category}'. "
                f"Valid categories: {', '.join(sorted(VALID_CATEGORIES))}"
            ),
        )

    try:
        query = (
            supabase.table("cultural_notes")
            .select("*", count="exact")
            .eq("cefr_level", cefr_level.value)
        )

        if category:
            query = query.eq("category", category)

        query = (
            query.range(offset, offset + limit - 1)
            .order("created_at", desc=True)
        )

        result = await query.execute()

        rows = result.data or []
        total = (
            result.count
            if result.count is not None
            else len(rows)
        )

        notes = []
        for row in rows:
            content_es = row.get("content_es", "")
            preview = (
                content_es[:150].rsplit(" ", 1)[0] + "..."
                if len(content_es) > 150
                else content_es
            )
            vocab_ids = row.get("vocabulary_ids") or []

            notes.append(
                CulturalNotePreview(
                    id=row["id"],
                    cefr_level=row["cefr_level"],
                    title_es=row["title_es"],
                    title_fr=row["title_fr"],
                    category=row["category"],
                    preview_es=preview,
                    vocabulary_count=len(vocab_ids),
                    reviewed=row.get("reviewed", False),
                )
            )

        return {
            "data": CulturalNoteListResponse(
                notes=notes,
                total=total,
            )
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Failed to list cultural notes")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve cultural notes.",
        ) from exc


# ---------------------------------------------------------------------------
# GET /notes/{id} -- Get note detail
# ---------------------------------------------------------------------------


@router.get(
    "/notes/{note_id}",
    response_model=dict[str, CulturalNoteDetail],
)
async def get_cultural_note(
    request: Request,
    note_id: UUID,
    user: UserInfo = Depends(get_current_user),
) -> dict[str, Any]:
    """Get the full content of a cultural note with vocabulary references."""
    supabase = _get_supabase(request)

    try:
        result = await (
            supabase.table("cultural_notes")
            .select("*")
            .eq("id", str(note_id))
            .execute()
        )

        if not result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Cultural note {note_id} not found.",
            )

        row = result.data[0]
        vocab_ids: list[str] = row.get("vocabulary_ids") or []

        # Fetch linked vocabulary items
        vocabulary: list[VocabularyRef] = []
        if vocab_ids:
            vocab_result = await (
                supabase.table("vocabulary_items")
                .select("id, french_text, spanish_translation")
                .in_("id", vocab_ids)
                .execute()
            )

            # Check which items are in the user's review queue
            user_progress = await (
                supabase.table("vocabulary_progress")
                .select("vocabulary_item_id")
                .eq("user_id", user.id)
                .execute()
            )
            in_queue = {
                r["vocabulary_item_id"]
                for r in (user_progress.data or [])
            }

            for v in vocab_result.data or []:
                vocabulary.append(
                    VocabularyRef(
                        id=v["id"],
                        french_text=v["french_text"],
                        spanish_translation=v["spanish_translation"],
                        in_user_review_queue=v["id"] in in_queue,
                    )
                )

        return {
            "data": CulturalNoteDetail(
                id=row["id"],
                cefr_level=row["cefr_level"],
                title_es=row["title_es"],
                title_fr=row["title_fr"],
                content_fr=row["content_fr"],
                content_es=row["content_es"],
                vocabulary=vocabulary,
                category=row["category"],
                is_generated=row.get("is_generated", True),
                reviewed=row.get("reviewed", False),
                created_at=row["created_at"],
            )
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Failed to get cultural note")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve cultural note.",
        ) from exc


# ---------------------------------------------------------------------------
# POST /notes/{id}/vocabulary/{vid}/add -- Add vocab to SRS
# ---------------------------------------------------------------------------


@router.post(
    "/notes/{note_id}/vocabulary/{vocab_id}/add",
    response_model=dict[str, AddVocabularyResponse],
    status_code=status.HTTP_201_CREATED,
)
async def add_vocabulary_to_srs(
    request: Request,
    note_id: UUID,
    vocab_id: UUID,
    user: UserInfo = Depends(get_current_user),
) -> dict[str, Any]:
    """Add a vocabulary item from a cultural note to the user's SRS queue."""
    supabase = _get_supabase(request)

    try:
        # Verify note exists and contains the vocabulary ID
        note_result = await (
            supabase.table("cultural_notes")
            .select("vocabulary_ids")
            .eq("id", str(note_id))
            .execute()
        )

        if not note_result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Cultural note {note_id} not found.",
            )

        # Verify vocabulary item exists
        vocab_result = await (
            supabase.table("vocabulary_items")
            .select("id")
            .eq("id", str(vocab_id))
            .execute()
        )

        if not vocab_result.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Vocabulary item {vocab_id} not found.",
            )

        # Check if already in user's review queue
        existing = await (
            supabase.table("vocabulary_progress")
            .select("id")
            .eq("user_id", user.id)
            .eq("vocabulary_item_id", str(vocab_id))
            .execute()
        )

        if existing.data:
            # Already in queue
            return {
                "data": AddVocabularyResponse(
                    vocabulary_item_id=vocab_id,
                    added_to_review=False,
                    first_review_date=datetime.now(UTC).isoformat(),
                )
            }

        # Create initial FSRS state and add to review queue
        initial_state = FSRSScheduler.initial_state()
        first_review = datetime.now(UTC) + timedelta(minutes=10)

        await (
            supabase.table("vocabulary_progress")
            .insert({
                "user_id": user.id,
                "vocabulary_item_id": str(vocab_id),
                "fsrs_stability": initial_state.stability,
                "fsrs_difficulty": initial_state.difficulty,
                "fsrs_due_date": first_review.isoformat(),
                "fsrs_interval": initial_state.scheduled_days,
                "review_count": 0,
                "correct_count": 0,
                "last_review_rating": 0,
                "last_reviewed_at": datetime.now(UTC).isoformat(),
            })
            .execute()
        )

        return {
            "data": AddVocabularyResponse(
                vocabulary_item_id=vocab_id,
                added_to_review=True,
                first_review_date=first_review.isoformat(),
            )
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Failed to add vocabulary to SRS")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to add vocabulary to review queue.",
        ) from exc


# ---------------------------------------------------------------------------
# POST /generate -- Trigger async cultural content generation
# ---------------------------------------------------------------------------


@router.post(
    "/generate",
    response_model=dict[str, GenerateResponse],
    status_code=status.HTTP_202_ACCEPTED,
)
async def generate_cultural_content(
    request: Request,
    body: GenerateRequest,
    user: UserInfo = Depends(get_current_user),
) -> dict[str, Any]:
    """Trigger asynchronous generation of a cultural note via Cloud Tasks."""
    supabase = _get_supabase(request)

    if body.category not in VALID_CATEGORIES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                f"Invalid category '{body.category}'. "
                f"Valid categories: {', '.join(sorted(VALID_CATEGORIES))}"
            ),
        )

    try:
        # Insert a pending async job for the worker
        import uuid

        generation_id = str(uuid.uuid4())

        payload = {
            "cefr_level": body.cefr_level.value,
            "category": body.category,
            "topic_hint": body.topic_hint,
            "align_with_vocabulary": [
                str(vid) for vid in body.align_with_vocabulary
            ],
            "generation_id": generation_id,
        }

        # Try to insert into async_jobs table for worker processing
        try:
            await (
                supabase.table("async_jobs")
                .insert({
                    "job_type": "cultural_content",
                    "status": "pending",
                    "payload": payload,
                    "user_id": user.id,
                })
                .execute()
            )
        except Exception:
            logger.warning(
                "async_jobs table not available, job will need manual processing"
            )

        return {
            "data": GenerateResponse(
                generation_id=generation_id,
                status="pending",
                ai_platform="gemini",
                estimated_completion_seconds=15,
            )
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Failed to trigger cultural content generation")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to trigger content generation.",
        ) from exc

File: web/src/services/cultural.ts
import { apiClient } from "./api";
import type { CEFRLevel } from "@/types/cefr";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface CulturalNotePreview {
  id: string;
  cefr_level: CEFRLevel;
  title_es: string;
  title_fr: string;
  category: string;
  preview_es: string;
  vocabulary_count: number;
  reviewed: boolean;
}

export interface CulturalNoteListResponse {
  notes: CulturalNotePreview[];
  total: number;
}

export interface VocabularyRef {
  id: string;
  french_text: string;
  spanish_translation: string;
  in_user_review_queue: boolean;
}

export interface CulturalNoteDetail {
  id: string;
  cefr_level: CEFRLevel;
  title_es: string;
  title_fr: string;
  content_fr: string;
  content_es: string;
  vocabulary: VocabularyRef[];
  category: string;
  is_generated: boolean;
  reviewed: boolean;
  created_at: string;
}

export interface AddVocabularyResponse {
  vocabulary_item_id: string;
  added_to_review: boolean;
  first_review_date: string;
}

export interface GenerateRequest {
  cefr_level: CEFRLevel;
  category: string;
  topic_hint?: string;
  align_with_vocabulary?: string[];
}

export interface GenerateResponse {
  generation_id: string;
  status: string;
  ai_platform: string;
  estimated_completion_seconds: number;
}

// ---------------------------------------------------------------------------
// Cultural category metadata
// ---------------------------------------------------------------------------

export type CulturalCategory =
  | "history"
  | "neighborhoods"
  | "etiquette"
  | "cuisine"
  | "daily_life";

export const CULTURAL_CATEGORIES: {
  key: CulturalCategory;
  labelEs: string;
  icon: string;
}[] = [
  { key: "history", labelEs: "Historia", icon: "M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0118 0z" },
  { key: "neighborhoods", labelEs: "Barrios", icon: "M19 21V5a2 2 0 00-2-2H7a2 2 0 00-2 2v16m14 0h2m-2 0h-5m-9 0H3m2 0h5M9 7h1m-1 4h1m4-4h1m-1 4h1m-5 10v-5a1 1 0 011-1h2a1 1 0 011 1v5m-4 0h4" },
  { key: "etiquette", labelEs: "Etiqueta", icon: "M17 20h5v-2a3 3 0 00-5.356-1.857M17 20H7m10 0v-2c0-.656-.126-1.283-.356-1.857M7 20H2v-2a3 3 0 015.356-1.857M7 20v-2c0-.656.126-1.283.356-1.857m0 0a5.002 5.002 0 019.288 0M15 7a3 3 0 11-6 0 3 3 0 016 0zm6 3a2 2 0 11-4 0 2 2 0 014 0zM7 10a2 2 0 11-4 0 2 2 0 014 0z" },
  { key: "cuisine", labelEs: "Gastronomia", icon: "M3 3h2l.4 2M7 13h10l4-8H5.4M7 13L5.4 5M7 13l-2.293 2.293c-.63.63-.184 1.707.707 1.707H17m0 0a2 2 0 100 4 2 2 0 000-4zm-8 2a2 2 0 100 4 2 2 0 000-4z" },
  { key: "daily_life", labelEs: "Vida cotidiana", icon: "M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-4 0a1 1 0 01-1-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 01-1 1" },
];

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/**
 * Fetch a paginated list of cultural notes, filtered by CEFR level and optional category.
 */
export async function getCulturalNotes(params: {
  cefr_level: CEFRLevel;
  category?: CulturalCategory;
  limit?: number;
  offset?: number;
}): Promise<{ data: CulturalNoteListResponse }> {
  const query = new URLSearchParams();
  query.set("cefr_level", params.cefr_level);
  if (params.category) query.set("category", params.category);
  if (params.limit) query.set("limit", String(params.limit));
  if (params.offset) query.set("offset", String(params.offset));
  return apiClient(`/cultural/notes?${query}`);
}

/**
 * Fetch the full content of a cultural note by ID.
 */
export async function getCulturalNote(
  id: string,
): Promise<{ data: CulturalNoteDetail }> {
  return apiClient(`/cultural/notes/${id}`);
}

/**
 * Add a vocabulary item from a cultural note to the user's SRS review queue.
 */
export async function addVocabularyFromNote(
  noteId: string,
  vocabId: string,
): Promise<{ data: AddVocabularyResponse }> {
  return apiClient(`/cultural/notes/${noteId}/vocabulary/${vocabId}/add`, {
    method: "POST",
  });
}

/**
 * Trigger asynchronous generation of a new cultural note.
 */
export async function generateCulturalContent(
  params: GenerateRequest,
): Promise<{ data: GenerateResponse }> {
  return apiClient("/cultural/generate", {
    method: "POST",
    body: JSON.stringify(params),
  });
}
